File: model-value/src/main/java/com/bnorm/barkeep/model/IngredientValue.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model;

import org.jetbrains.annotations.NotNull;

import com.google.auto.value.AutoValue;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

@AutoValue
public abstract class IngredientValue implements Ingredient {

  public static Builder builder() {
    return new AutoValue_IngredientValue.Builder();
  }

  public static JsonAdapter<IngredientValue> jsonAdapter(Moshi moshi) {
    return new AutoValue_IngredientValue.MoshiJsonAdapter(moshi);
  }

  @Override
  public int compareTo(@NotNull Ingredient other) {
    return Ingredient.Companion.getCOMPARATOR().compare(this, other);
  }

  @AutoValue.Builder
  public abstract static class Builder {

    public abstract IngredientValue build();

    public abstract Builder setId(Long id);

    public abstract Builder setTitle(String title);

    public abstract Builder setParent(Ingredient ingredient);
  }
}

File: model-value/src/main/kotlin/com/bnorm/barkeep/model/IngredientValue.kt
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.squareup.moshi.FromJson
import com.squareup.moshi.ToJson

data class IngredientSpecValue(override val title: String?,
                               override val parent: Ingredient? = null) : IngredientSpec

data class IngredientValue(override val id: Long = -1,
                           override val title: String?,
                           override val parent: Ingredient? = null) : Ingredient

object IngredientValueAdapter {
  @ToJson fun toJson(ingredient: IngredientSpec): IngredientSpecValue {
    return IngredientSpecValue(ingredient.title, ingredient.parent)
  }

  @FromJson fun fromJson(json: IngredientSpecValue): IngredientSpec {
    return json
  }

  @ToJson fun toJson(ingredient: Ingredient): IngredientValue {
    return IngredientValue(ingredient.id, ingredient.title, ingredient.parent)
  }

  @FromJson fun fromJson(json: IngredientValue): Ingredient {
    return json
  }
}
